src/main-container/private/ActiveMainProvider.jsx

```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import uuidv4 from 'uuid/v4';
+
+import NavigationItemContext from '../../navigation-item/NavigationItemContext';
+
+import ActiveMainContext from '../ActiveMainContext';
+import ActiveMainRegistrationContext from './ActiveMainRegistrationContext';
+
+const propTypes = {
+  children: PropTypes.node,
+};
+
+const ActiveMainProvider = ({ children }) => {
+  const navigationItem = React.useContext(NavigationItemContext);
+  const activeMainRegistration = React.useContext(ActiveMainRegistrationContext);
+  const unregisterActiveMainRef = React.useRef();
+
+  const [state, dispatch] = React.useReducer((currentState, action) => {
+    if (action.type === 'register') {
+      return {
+        registrationId: action.registrationId,
+        activeMain: {
+          label: action.label,
+          metaData: action.metaData,
+        },
+      };
+    }
+
+    if (action.type === 'unregister') {
+      if (currentState.registrationId === action.registrationId) {
+        return {
+          registrationId: undefined,
+          activeMainPage: undefined,
+        };
+      }
+    }
+
+    return currentState;
+  }, { registrationId: undefined, activeMain: undefined });
+
+  React.useEffect(() => {
+    if (!activeMainRegistration) {
+      return;
+    }
+
+    // If an ancestor provider exists, we need to forward the active main info
+    // if the provider exists in the active navigation branch. It is
+    // otherwise unregistered, if necessary, as we do not want potentially stale
+    // information living above this provider level.
+    if (navigationItem.isActive) {
+      unregisterActiveMainRef.current = activeMainRegistration.register(state.activeMain);
+    } else if (unregisterActiveMainRef.current) {
+      unregisterActiveMainRef.current();
+      unregisterActiveMainRef.current = undefined;
+    }
+  }, [state.activeMain, navigationItem.isActive, activeMainRegistration]);
+
+  React.useEffect(() => () => {
+    if (unregisterActiveMainRef.current) {
+      unregisterActiveMainRef.current();
+      unregisterActiveMainRef.current = undefined;
+    }
+  }, []);
+
+  const activeMainRegistrationContextValue = React.useMemo(() => ({
+    register: (registrationData) => {
+      if (!registrationData) {
+        return undefined;
+      }
+
+      const { label, metaData } = registrationData;
+      const registrationId = uuidv4();
```

Comment:
Multiple main sources can be writing to this single provider, and the order of their registrations isn't guaranteed to be deterministic. So I assign an identifier to each registration and check it above prior to performing any unregistration, ensuring that main elements can execute their unregistration logic logically as part of their lifecycle without worrying about damaging registration data from other sources.

Reply:
This might be more deterministic in the future with React 17 based on some PRs I've seen: https://github.com/facebook/react/pull/17925

Reply:
we could add what you've outlined as a comment/todo